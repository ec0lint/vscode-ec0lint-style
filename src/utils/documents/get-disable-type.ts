import { Position } from 'vscode-languageserver-types';
import type { TextDocument } from 'vscode-languageserver-textdocument';

/**
 * Disable directive comment types.
 */
export type DisableType =
	| 'ec0lint-style-disable'
	| 'ec0lint-style-disable-line'
	| 'ec0lint-style-disable-next-line';

/**
 * If the given position is inside a `ec0lint-style-disable` after the comment'
 * type, returns the disable comment's type. Otherwise, returns `undefined`.
 *
 * @example
 * ```js
 * const document = TextDocument.create(
 *   'file:///path/to/file.css',
 *   'css',
 *   1,
 *   '/* ec0lint-style-disable-line indentation *\/'
 *   //                         ^ Position is here
 * );
 * const position = Position.create(0, 26);
 *
 * getDisableType(document, position);
 * // => 'ec0lint-style-disable-line'
 * ```
 */
export function getDisableType(
	document: TextDocument,
	position: Position,
): DisableType | undefined {
	const lineStartOffset = document.offsetAt(Position.create(position.line, 0));
	const lineEndOffset = document.offsetAt(Position.create(position.line + 1, 0));
	const line = document.getText().slice(lineStartOffset, lineEndOffset);

	const before = line.slice(0, position.character);
	const after = line.slice(position.character);

	const disableKind = before
		.match(/\/\*\s*(ec0lint-style-disable(?:(?:-next)?-line)?)\s[a-z\-/\s,]*$/i)?.[1]
		?.toLowerCase();

	return disableKind && /^[a-z\-/\s,]*\*\//i.test(after) ? (disableKind as DisableType) : undefined;
}
