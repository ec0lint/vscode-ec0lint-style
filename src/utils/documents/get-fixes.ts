import { createTextEdits } from './create-text-edits';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import type { Ec0lintRunner, RunnerOptions } from '../ec0lint-style';

/**
 * Runs Ec0lint and returns fix text edits for the given document.
 * @param runner The Ec0lint runner.
 * @param document The document to get fixes for.
 * @param linterOptions Linter options to use.
 * @param runnerOptions The runner options.
 */
export async function getFixes(
	runner: Ec0lintRunner,
	document: TextDocument,
	linterOptions: Ec0lintStyle.LinterOptions = {},
	runnerOptions: RunnerOptions = {},
): Promise<TextEdit[]> {
	const result = await runner.lintDocument(
		document,
		{ ...linterOptions, fix: true },
		runnerOptions,
	);

	return typeof result.output === 'string' ? createTextEdits(document, result.output) : [];
}
