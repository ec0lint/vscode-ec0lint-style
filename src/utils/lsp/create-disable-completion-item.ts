import {
	CompletionItem,
	CompletionItemKind,
	InsertTextFormat,
	MarkupKind,
} from 'vscode-languageserver-types';
import type { DisableType } from '../documents';

/**
 * Creates a disable completion item for the given disable type. Uses the given rule if one is
 * provided, otherwise uses a placeholder.
 */
export function createDisableCompletionItem(disableType: DisableType, rule = ''): CompletionItem {
	const item = CompletionItem.create(disableType);

	item.kind = CompletionItemKind.Snippet;
	item.insertTextFormat = InsertTextFormat.Snippet;

	if (disableType === 'ec0lint-style-disable') {
		item.insertText = `/* ec0lint-style-disable \${0:${rule || 'rule'}} */\n/* ec0lint-style-enable \${0:${
			rule || 'rule'
		}} */`;
		item.detail =
			'Turn off all ec0lint-style or individual rules, after which you do not need to re-enable ec0lint-style. (ec0lint-style)';
		item.documentation = {
			kind: MarkupKind.Markdown,
			value: `\`\`\`css\n/* ec0lint-style-disable ${rule || 'rule'} */\n/* ec0lint-style-enable ${
				rule || 'rule'
			} */\n\`\`\``,
		};
	} else {
		item.insertText = `/* ${disableType} \${0:${rule || 'rule'}} */`;
		item.detail =
			disableType === 'ec0lint-style-disable-line'
				? 'Turn off ec0lint-style rules for individual lines only, after which you do not need to explicitly re-enable them. (ec0lint-style)'
				: 'Turn off ec0lint-style rules for the next line only, after which you do not need to explicitly re-enable them. (ec0lint-style)';
		item.documentation = {
			kind: MarkupKind.Markdown,
			value: `\`\`\`css\n/* ${disableType} ${rule || 'rule'} */\n\`\`\``,
		};
	}

	return item;
}
