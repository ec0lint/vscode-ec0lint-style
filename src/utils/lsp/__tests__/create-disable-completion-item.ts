import { DisableType } from '../../documents';
import { createDisableCompletionItem } from '../create-disable-completion-item';

describe('createDisableCompletionItem', () => {
	test('should create a ec0lint-style-disable completion item', () => {
		const result = createDisableCompletionItem('ec0lint-style-disable' as DisableType);

		expect(result).toMatchSnapshot();
	});

	test('should create a ec0lint-style-disable completion item for a specific rule', () => {
		const result = createDisableCompletionItem('ec0lint-style-disable' as DisableType, 'indentation');

		expect(result).toMatchSnapshot();
	});

	test('should create a ec0lint-style-disable-line completion item', () => {
		const result = createDisableCompletionItem('ec0lint-style-disable-line' as DisableType);

		expect(result).toMatchSnapshot();
	});

	test('should create a ec0lint-style-disable-line completion item for a specific rule', () => {
		const result = createDisableCompletionItem('ec0lint-style-disable-line' as DisableType, 'indentation');

		expect(result).toMatchSnapshot();
	});

	test('should create a ec0lint-style-disable-next-line completion item', () => {
		const result = createDisableCompletionItem('ec0lint-style-disable-next-line' as DisableType);

		expect(result).toMatchSnapshot();
	});

	test('should create a ec0lint-style-disable-next-line completion item for a specific rule', () => {
		const result = createDisableCompletionItem('ec0lint-style-disable-next-line' as DisableType, 'indentation');

		expect(result).toMatchSnapshot();
	});
});
