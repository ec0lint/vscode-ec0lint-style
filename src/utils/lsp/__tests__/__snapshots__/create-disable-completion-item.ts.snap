// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDisableCompletionItem should create a ec0lint-style-disable completion item 1`] = `
Object {
  "detail": "Turn off all ec0lint-style or individual rules, after which you do not need to re-enable ec0lint-style. (ec0lint-style)",
  "documentation": Object {
    "kind": "markdown",
    "value": "\`\`\`css
/* ec0lint-style-disable rule */
/* ec0lint-style-enable rule */
\`\`\`",
  },
  "insertText": "/* ec0lint-style-disable \${0:rule} */
/* ec0lint-style-enable \${0:rule} */",
  "insertTextFormat": 2,
  "kind": 15,
  "label": "ec0lint-style-disable",
}
`;

exports[`createDisableCompletionItem should create a ec0lint-style-disable completion item for a specific rule 1`] = `
Object {
  "detail": "Turn off all ec0lint-style or individual rules, after which you do not need to re-enable ec0lint-style. (ec0lint-style)",
  "documentation": Object {
    "kind": "markdown",
    "value": "\`\`\`css
/* ec0lint-style-disable indentation */
/* ec0lint-style-enable indentation */
\`\`\`",
  },
  "insertText": "/* ec0lint-style-disable \${0:indentation} */
/* ec0lint-style-enable \${0:indentation} */",
  "insertTextFormat": 2,
  "kind": 15,
  "label": "ec0lint-style-disable",
}
`;

exports[`createDisableCompletionItem should create a ec0lint-style-disable-line completion item 1`] = `
Object {
  "detail": "Turn off ec0lint-style rules for individual lines only, after which you do not need to explicitly re-enable them. (ec0lint-style)",
  "documentation": Object {
    "kind": "markdown",
    "value": "\`\`\`css
/* ec0lint-style-disable-line rule */
\`\`\`",
  },
  "insertText": "/* ec0lint-style-disable-line \${0:rule} */",
  "insertTextFormat": 2,
  "kind": 15,
  "label": "ec0lint-style-disable-line",
}
`;

exports[`createDisableCompletionItem should create a ec0lint-style-disable-line completion item for a specific rule 1`] = `
Object {
  "detail": "Turn off ec0lint-style rules for individual lines only, after which you do not need to explicitly re-enable them. (ec0lint-style)",
  "documentation": Object {
    "kind": "markdown",
    "value": "\`\`\`css
/* ec0lint-style-disable-line indentation */
\`\`\`",
  },
  "insertText": "/* ec0lint-style-disable-line \${0:indentation} */",
  "insertTextFormat": 2,
  "kind": 15,
  "label": "ec0lint-style-disable-line",
}
`;

exports[`createDisableCompletionItem should create a ec0lint-style-disable-next-line completion item 1`] = `
Object {
  "detail": "Turn off ec0lint-style rules for the next line only, after which you do not need to explicitly re-enable them. (ec0lint-style)",
  "documentation": Object {
    "kind": "markdown",
    "value": "\`\`\`css
/* ec0lint-style-disable-next-line rule */
\`\`\`",
  },
  "insertText": "/* ec0lint-style-disable-next-line \${0:rule} */",
  "insertTextFormat": 2,
  "kind": 15,
  "label": "ec0lint-style-disable-next-line",
}
`;

exports[`createDisableCompletionItem should create a ec0lint-style-disable-next-line completion item for a specific rule 1`] = `
Object {
  "detail": "Turn off ec0lint-style rules for the next line only, after which you do not need to explicitly re-enable them. (ec0lint-style)",
  "documentation": Object {
    "kind": "markdown",
    "value": "\`\`\`css
/* ec0lint-style-disable-next-line indentation */
\`\`\`",
  },
  "insertText": "/* ec0lint-style-disable-next-line \${0:indentation} */",
  "insertTextFormat": 2,
  "kind": 15,
  "label": "ec0lint-style-disable-next-line",
}
`;
